### ISSUE TITLE
Remove mutability of self reference in getter methods of Exhange trait

### ISSUE THREAD

**Issue by mankinskin** [nash-io/openlimits#75]
"get" methods should not require mutable access to `self`.  I know this was probably done to provide more flexibility to trait users, but this breaks the common principle of "get" being a reading operation, and requires trait users to ensure mutable access all the way through the code that uses `Exchange`.
```rust
#[async_trait]
pub trait Exchange {
    type OrderIdType;
    type TradeIdType;
    type PaginationType;
    async fn refresh_market_info(&mut self) -> Result<()>;
    async fn order_book(&mut self, req: &OrderBookRequest) -> Result<OrderBookResponse>;
    async fn limit_buy(&mut self, req: &OpenLimitOrderRequest) -> Result<Order<Self::OrderIdType>>;
    async fn limit_sell(&mut self, req: &OpenLimitOrderRequest)
        -> Result<Order<Self::OrderIdType>>;
    async fn market_buy(
        &mut self,
        req: &OpenMarketOrderRequest,
    ) -> Result<Order<Self::OrderIdType>>;
    async fn market_sell(
        &mut self,
        req: &OpenMarketOrderRequest,
    ) -> Result<Order<Self::OrderIdType>>;
    async fn cancel_order(
        &mut self,
        req: &CancelOrderRequest<Self::OrderIdType>,
    ) -> Result<OrderCanceled<Self::OrderIdType>>;
    async fn cancel_all_orders(
        &mut self,
        req: &CancelAllOrdersRequest,
    ) -> Result<Vec<OrderCanceled<Self::OrderIdType>>>;
    async fn get_all_open_orders(&mut self) -> Result<Vec<Order<Self::OrderIdType>>>;
    async fn get_order_history(
        &mut self,
        req: &GetOrderHistoryRequest<Self::PaginationType>,
    ) -> Result<Vec<Order<Self::OrderIdType>>>;
    async fn get_account_balances(
        &mut self,
        paginator: Option<&Paginator<Self::PaginationType>>,
    ) -> Result<Vec<Balance>>;
    async fn get_trade_history(
        &mut self,
        req: &TradeHistoryRequest<Self::OrderIdType, Self::PaginationType>,
    ) -> Result<Vec<Trade<Self::TradeIdType, Self::OrderIdType>>>;
    async fn get_price_ticker(&mut self, req: &GetPriceTickerRequest) -> Result<Ticker>;
    async fn get_historic_rates(
        &mut self,
        req: &GetHistoricRatesRequest<Self::PaginationType>,
    ) -> Result<Vec<Candle>>;
    async fn get_order(
        &mut self,
        req: &GetOrderRequest<Self::OrderIdType>,
    ) -> Result<Order<Self::OrderIdType>>;
}
```

**Comment by acatinon** [comment]
I agree, I wrote a similar message on slack yesterday :)

**Comment by mankinskin** [comment]
@steffenix (https://github.com/nash-io/openlimits/pull/76)
> I would prefere using non mut version of mut but this will prevent nash from using openlimits.
> Nash uses a pool of MPC keys that needs to be mutated when doing some signatures.
> 
> @Ejhfast that would be good to opensource the nash rust sdk so people start seeing why we need to use mut self.

Another option would be to implement mut and non-mut versions. Something like
```rust
async fn get_price_ticker(&self, req: &GetPriceTickerRequest) -> Result<Ticker> {
    compile_error!("Not implemented");
}
async fn mut_get_price_ticker(&mut self, req: &GetPriceTickerRequest) -> Result<Ticker> {
    self.get_price_ticker(req).await
}
```
This would allow trait users to define both versions, the default being the immutable version. If you only use the mutable version you can override `mut_get_price_ticker`, and the immutable version will cause a `compile_error` when used.

**Comment by acatinon** [comment]
This option should be implemented in last resort, as it will make openLimits quite tedious to use and implement new exchange.

Maybe we can sort this out using interior mutability ? Indeed it will help to opensource the nash rust sdk.

**Comment by mankinskin** [comment]
I am not part of the nash project, I just use openlimits as an API to multiple exchanges. The problem with requiring `mut`, is that this would require me to use `mut` everywhere where I just want to do a reading operation, which often requires extra code to handle write synchronization and is not very sound anyways.

> it will make openLimits quite tedious to use and implement new exchanges

I don't think it will make the trait any more tedious to use. You can just implement any of the two, or both of them. If you try to use one that you didn't define, the `compile_error!` macro will be encountered and you get a compile-time error.

**Comment by Ejhfast** [comment]
We'll try to get the nash sdk out soon. The nash sdk client requires `&mut
self` right now even for getters because of some stateful bookkeeping it
needs to do when maintaining a WS connection with Absinthe. I haven't
looked enough at openlimits enough to suggest any immediate work arounds,
but it's likely possible for me to abstract this away on the sdk side if
it's causing big enough headaches for others. Will take another look at it

**Comment by Ejhfast** [comment]
It turns out it was pretty easy to update the nash sdk with an interior mutability pattern, so this is unblocked now.
